src/core/ddsc/src/dds_reader.c

```diff
@@ -676,23 +673,43 @@ static dds_entity_t dds_create_reader_int (dds_entity_t participant_or_subscribe
     // shared memory but limit the queueCapacity accordingly (otherwise iceoryx
     // emits a warning and limits it itself)
 
-    if (rqos->history.depth <= max_sub_queue_capacity) {
+    if ((uint32_t) rqos->history.depth <= max_sub_queue_capacity) {
       opts.queueCapacity = (uint64_t)rqos->history.depth;
     } else {
-      opts.queueCapacity = (uint64_t)max_sub_queue_capacity;
+      opts.queueCapacity = max_sub_queue_capacity;
     }
 
     // NB: since currently we only support the volatile reader case we will
     // never request historical data
     opts.historyRequest = 0;
-    
-    rd->m_iox_sub = iox_sub_init(&rd->m_iox_sub_stor.storage, gv->config.iceoryx_service, rd->m_topic->m_stype->type_name, rd->m_topic->m_name, &opts);
-    shm_monitor_attach_reader(&rd->m_entity.m_domain->m_shm_monitor, rd);
+
+    // NB: This may fail due to icoeryx being out of internal resources for subsribers.
+    //     In this case terminate is called by iox_sub_init.
+    //     it is currently (iceoryx 2.0 and lower) not possible to change this to
+    //     e.g. return a nullptr and handle the error here.
+    iox_sub_storage_t s; // storage is ignored internally now but we cannot pass a nullptr
+    rd->m_iox_sub = iox_sub_init(&s, gv->config.iceoryx_service, rd->m_topic->m_stype->type_name, rd->m_topic->m_name, &opts);
+
+    // NB: Due to some storage paradigm change of iceoryx structs
+    // we now have a pointer 8 bytes before m_iox_sub
+    // We use this address to store a pointer to the context.
+    iox_sub_context_t **context = get_context_ptr(rd->m_iox_sub);
+    *context = &rd->m_iox_sub_context;
+
+    rc = shm_monitor_attach_reader(&rd->m_entity.m_domain->m_shm_monitor, rd);
+
+    if (rc != DDS_RETCODE_OK) {
+      // we fail if we cannot attach to the listener (as we would get no data)
+      iox_sub_deinit(rd->m_iox_sub);
+      rd->m_iox_sub = NULL;
+      DDS_CLOG (DDS_LC_SHM, &rd->m_entity.m_domain->gv.logconfig, "Failed to attach iox subscriber to iox listener\n");
+      goto err_bad_qos; // we need to clean up everything
```

Comment:
Not enough, deinit partially created entity? 

Complete creation and delete it?

Reply:
FIXME comment that indicates we need to extend the cleanup.